### PR TITLE
Transaction: Tests and fix for building simple Segwit transactions (without wallet or network)

### PR DESCRIPTION
**Pull request by msgilligan**
This PR contains a test that reproduces issue #2120 and also provides a possible fix.

* Commit fd49bf263370d91e9bba15c182eadc73ee28bfef adds two tests: one for Legacy addresses that works and one for Segwit addresses that fails (reproduces Issue #2120)
* Commit b926cb8c2b0d82c1473966e2f6ec7a80081796d2 provides a possible fix.

@schildbach if this looks good to you, let me know and I'll add JavaDoc and squash.

### CODE REVIEW

File: core/src/test/java/org/bitcoinj/core/TransactionTest.java [hunk 1]
```diff
@@ -168,6 +168,43 @@ public void testIsMatureReturnsFalseIfTransactionIsCoinbaseAndConfidenceTypeIsNo
...
+        byte[] rawTx = tx.bitcoinSerialize();
+
+        assertNotNull(rawTx);
+    }
+
+    @Test
+    public void testBuildingSimpleP2WPKH() {
+        final Address toAddr = Address.fromKey(TESTNET, new ECKey(), Script.ScriptType.P2WPKH);
+        final Sha256Hash utxo_id = Sha256Hash.wrap("81b4c832d70cb56ff957589752eb4125a4cab78a25a8fc52d6a09e5bd4404d48");
+        final Coin inAmount = Coin.ofSat(91234);
```
Comment (schildbach) [inline core/src/test/java/org/bitcoinj/core/TransactionTest.java:193]: Seems unused?

File: core/src/test/java/org/bitcoinj/core/TransactionTest.java [hunk 2]
```diff
@@ -198,7 +198,7 @@ public void testBuildingSimpleP2WPKH() {
         Transaction tx = new Transaction(TESTNET);
         TransactionOutPoint outPoint = new TransactionOutPoint(TESTNET, 0, utxo_id);
         tx.addOutput(outAmount, toAddr);
-        tx.addSignedInput(outPoint, ScriptBuilder.createOutputScript(fromAddress), fromKey);
```
Comment (schildbach) [inline core/src/test/java/org/bitcoinj/core/TransactionTest.java]: Ok, now I understand (-:

**Comment by schildbach** [comment]
Ok, then I'd vote for being harsh on 0.17 but being lenient on the 0.16 backport.

**Comment by msgilligan** [comment]
> Ok, then I'd vote for being harsh on 0.17 but being lenient on the 0.16 backport.

OK, cool. I'll update this PR to be the 0.17 version and then we can create the backport after this is merged. Sound good?

**Comment by schildbach** [comment]
Sounds good.

**Comment by msgilligan** [comment]
OK, @schildbach I've deprecated two methods, added some checks and warning logs, and also made sure the JavaDoc for all the `addSignedInput` methods is "complete" (generates now warnings from the javadoc linter)

If this looks good, I can squash and force-push.

**Comment by schildbach** [comment]
Looks good!

Two remarks:

- Can we safely assume inputs to always have a value greater than 0? (Outputs can be 0, if you think of OP_RETURN)
- Should we aim for a consistent capitalization of the word segwit vs SegWit?

**Comment by msgilligan** [comment]
> Can we safely assume inputs to always have a value greater than 0?

I think so. How would we know for sure?


> Should we aim for a consistent capitalization of the word segwit vs SegWit?

Yes. What does the Bitcoin Core documentation use?

**Comment by schildbach** [comment]
The bitcoin repository:

```
$ grep -R segwit | wc -l
661
$ grep -R SegWit | wc -l
59
```

**Comment by schildbach** [comment]
The bitcoin.org website:

```
$ grep -R segwit | wc -l
224
$ grep -R SegWit | wc -l
55
```

**Comment by schildbach** [comment]
The BIPs repository:

```
$ grep -R segwit | wc -l
84
$ grep -R SegWit | wc -l
17
```

I think the small spelling wins.

**Comment by schildbach** [comment]
> I think so. How would we know for sure?

I think so as well, but I don't know how to be sure.

**Comment by schildbach** [comment]
I think because outputs with zero value are invalid (except for OP_RETURN), any input will have a non-zero value attached to it implicitely (OP_RETURN cannot be spent).

For unit testing, we typically use `Coin.COIN` or similar if the value is not important. I see no need to allow e.g. `Coin.ZERO` at this point.

A somewhat similar case may be fee. We allow zero fee in out API, and for unit testing it's indeed very comfortable, but today's network will likely not relay or mine it (though a miner might mine such a tx directly).

**Comment by schildbach** [comment]
Since no one came of with a reason to allow zero value, let's squash and merge, shall we?

**Comment by msgilligan** [comment]
@schildbach Squashed, force-merged, and (hopefully) ready-to-go!

**Comment by schildbach** [comment]
Merged.

**Pull request closed without merging**